### PR TITLE
ENH: option to plot data points on top of boxplot

### PR DESCRIPTION
**Pull request by mikessut**
- [x ] tests added / passed 
- [x ] passes `black pandas`
- [x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry - not sure how to do this. I guess I have to wait until pull request is complete.

This change is a a feature I often want, which is the ability to add raw data on top of a boxplot to get a better idea of data distribution.

### CODE REVIEW

File: pandas/plotting/_matplotlib/boxplot.py [hunk 1]
```diff
@@ -296,10 +297,18 @@ def maybe_color_bp(bp, **kwds):
...
+    def plot_group(keys, values, ax: "Axes", plot_points: bool = False):
         keys = [pprint_thing(x) for x in keys]
         values = [np.asarray(remove_na_arraylike(v), dtype=object) for v in values]
         bp = ax.boxplot(values, **kwds)
+
+        if plot_points:
+            # if plot_points
+            for x, ys in enumerate(values):
+                ax.plot(
+                    x + np.random.normal(size=len(ys)) * 0.04 + 1, ys, "ro", alpha=0.5
```
Comment (ivanovmg) [inline pandas/plotting/_matplotlib/boxplot.py:309]: Can you explain what this 0.04 and + 1 mean here?
I suggest to extract a function and use named variables there for clarity.

File: pandas/plotting/_matplotlib/boxplot.py [hunk 2]
```diff
@@ -296,10 +297,18 @@ def maybe_color_bp(bp, **kwds):
...
             setp(bp["caps"], color=colors[3], alpha=1)
 
-    def plot_group(keys, values, ax: "Axes"):
+    def plot_group(keys, values, ax: "Axes", plot_points: bool = False):
         keys = [pprint_thing(x) for x in keys]
         values = [np.asarray(remove_na_arraylike(v), dtype=object) for v in values]
         bp = ax.boxplot(values, **kwds)
+
+        if plot_points:
+            # if plot_points
```
Comment (ivanovmg) [inline pandas/plotting/_matplotlib/boxplot.py:306]: I think that the comment here is not required.

**Review by ivanovmg — CHANGES_REQUESTED** [review]
@mikessut thank you for the PR.

I guess that for this kind of PR it is necessary to show how the visualization works before/after,
since it is rather complicated to infer that from the code alone.
Can you add a visual example?

**Comment by mikessut** [comment]
You're right. That comment was just leftover from my hacking. I removed it.

Here's an example below.  It's inspired by a few examples here:
https://stackoverflow.com/questions/29779079/adding-a-scatter-of-points-to-a-boxplot-using-matplotlib
Use of seaborn's swarmplot on top of boxplot.
https://seaborn.pydata.org/generated/seaborn.boxplot.html


The .04 is the standard deviation of the noise added to the x value so that
the points don't end up on top of each other. The underlying x tick
apparently start at 1 (I can't claim to understand exactly how/why).
[image: image.png]

**Comment by mikessut** [comment]
Try that again to get example...
image

**Comment by ivanovmg** [comment]
I guess that the feedback from the core team on the plotting is necessary here.

It seems to me that three points require attention: 
 - Public API (new parameter)
 - Hard-coded color. Should it be optional? If so, then should it be selected from ``color``?
 - Hard-coded width of the distribution. Will it work for small and large (say 10) number of boxes? Would not they overlap?

**Comment by charlesdong1991** [comment]
IMHO, i don't think this should be implemented on pandas for now, this might be for matplotlib.

IIUC this implementation is adding points using ax.plot and this should be separated apart from boxplot, and the significance of boxplot is not plotting individual points but for plotting them through quartiles (more from a statistical perspective). 

So if users want to plot those points, they could use the code you have in PR and put it after df.boxplot(ax=ax), and so that users also have the flexibility to change color or shape or size of those points. Therefore, i would suggest creating an issue to collect some feedbacks first before diving into the detailed implementation (and hard-coded color and floats need to be reconsidered)

**Comment by mikessut** [comment]
FWIW, my longing for this feature comes from many years of using Minitab, which has a convenient option to show [individual data on boxplots](https://support.minitab.com/en-us/minitab/18/help-and-how-to/graphs/how-to/general-graph-options/display-options/boxplot/).  As a result, I often find myself wanting to be able to get a better visual idea of the data distribution.  

You are absolutely correct, that the data can be plotted by other means.  However, I haven't found those other means particularly intuitive and every time I want to show the underlying data, I find myself digging through stack overflow trying to remember how to do it.  The last time I found myself doing this, I decided I'd create the pull request.

Seaborn appears to have quite a few of the features that I'm looking for -- both boxplot and categorical scatter plots.  Maybe that's a better workflow for what I'm trying to accomplish. E.g.:
```
df = pd.DataFrame(np.random.randn(10, 3))
df2 = df.melt()

sns.boxplot(x='variable', y='value', data=df2)
sns.stripplot(x='variable', y='value', data=df2)
```

That get's pretty close, but certainly not as convenient as the following when trying to quickly inspect the data:
```
df = pd.DataFrame(np.random.randn(10, 3))
df.boxplot(plot_points=True)
```

**Comment by ivanovmg** [comment]
I think this is a reasonable suggestion, to make the inspection of the data easier.
However I agree that it is better to move this discussion to a dedicated enhancement-related issue and collect feedback from those concerned.

**Comment by charlesdong1991** [comment]
thanks for your comments!
 
As I said, the usage of boxplot is _not_ plotting individual points, but I do understand allowing this option might make the inspection easier in certain circumstances. However, there will be many issues to think of, especially if there are large amount of data points, I can imagine plotting individual points can be very messy (in your example there are 10, but i think 10000 is enough to make the visual not readable at all). And having this is also an API change, so definitely need to collect feedbacks first.

Could you please create an issue first for this? @mikessut

**Comment by mikessut** [comment]
Created issue:
https://github.com/pandas-dev/pandas/issues/37759

**Comment by MarcoGorelli** [comment]
closing, see https://github.com/pandas-dev/pandas/issues/37759#issuecomment-774693593

**Pull request closed without merging**